move and fix logic to update unbacked bindings

avikchaudhuri opened the pull request:
Summary:
Previously we were touching up unbacked bindings between Dynamo and AOTAutograd in strict export, but the logic had a bug: if an unbacked symint gets substituted by a backed symint, we would put the backed symint in the unbacked bindings (the check `is_symbol` was not enough here).

This PR fixes this logic, and moreover, moves it into the serializer instead, because we don't need this adjustment outside serde.

Test Plan: added test

 D68880766

pianpwk commented on torch/_export/serde/serialize.py:
```diff
@@ -603,9 +603,19 @@ def serialize_metadata(self, node: torch.fx.Node) -> dict[str, str]:
+            if new_unbacked_bindings:
+                ret["unbacked_bindings"] = ",".join(
+                    u.name for u in new_unbacked_bindings.keys()
+                )
```
Is it slightly weird that serialization is storing something different from what it's given? I liked where this was before...

avikchaudhuri replied:
I hear you. However in the current state we're recreating a shape env in serde, and we don't in the normal flow; running decompositions in the normal flow reuses shape env and doesn't need any adjustment, so I'd like to preserve behavior there and isolate these changes, which are really about shape env reconstruction, to serde. Does that make sense?

avikchaudhuri committed:
move and fix logic to update unbacked bindings (#146115)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/146115

Previously we were touching up unbacked bindings between Dynamo and AOTAutograd in strict export, but the logic had a bug: if an unbacked symint gets substituted by a backed symint, we would put the backed symint in the unbacked bindings (the check `is_symbol` was not enough here).

This PR fixes this logic, and moreover, moves it into the serializer instead, because we don't need this adjustment outside serde.

Test Plan: added test

Reviewed By: pianpwk

Differential Revision: D68880766

izaitsevfb commented:
@pytorchbot merge -f 'landed internally;

izaitsevfb commented:
@pytorchbot merge -f 'landed internally'

huydhn commented:
@pytorchbot revert -m 'This change has been reverted internally D69129334 but the OSS revert failed https://github.com/pytorch/pytorch/pull/146437' -c ghfirst
